cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowDispatch.qll

```diff
@@ -22,57 +23,125 @@ Function viableCallable(CallInstruction call) {
     strictcount(Function other | functionSignatureWithBody(qualifiedName, nparams, other)) = 1
   )
   or
-  // Rudimentary virtual dispatch support. It's essentially local data flow
-  // where the source is a derived-to-base conversion and the target is the
-  // qualifier of a call.
-  exists(Class derived, DataFlow::Node thisArgument |
-    nodeMayHaveClass(derived, thisArgument) and
-    overrideMayAffectCall(derived, thisArgument, _, result, call)
-  )
+  // Virtual dispatch
+  result = call.(VirtualDispatch::DataSensitiveCall).resolve()
 }
 
 /**
- * Holds if `call` is a virtual function call with qualifier `thisArgument` in
- * `enclosingFunction`, whose static target is overridden by
- * `overridingFunction` in `overridingClass`.
+ * Provides virtual dispatch support compatible with the original
+ * implementation of `semmle.code.cpp.security.TaintTracking`.
  */
-pragma[noinline]
-private predicate overrideMayAffectCall(
-  Class overridingClass, DataFlow::Node thisArgument, Function enclosingFunction,
-  MemberFunction overridingFunction, CallInstruction call
-) {
-  call.getEnclosingFunction() = enclosingFunction and
-  overridingFunction.getAnOverriddenFunction+() = call.getStaticCallTarget() and
-  overridingFunction.getDeclaringType() = overridingClass and
-  thisArgument = DataFlow::instructionNode(call.getThisArgument())
-}
+private module VirtualDispatch {
+  /** A call that may dispatch differently depending on the qualifier value. */
+  abstract class DataSensitiveCall extends DataFlowCall {
+    abstract DataFlow::Node getSrc();
```

Comment:
In a data-flow sense, it's more of a sink than a source. I think it's called _source_ in `security.TaintTracking` because the data flow is implemented backwards, from sink to source. I could rename it to `getDispatchValue` or something like that if it's helpful? The predicate gets the `x` in `x.virtualFunction()` or `x()`.

Reply:
I've added a commit to rename and document `getSrc()`.

Reply:
Thanks.  I think little changes like this can be a big benefit for people who are trying to understand the QL for the first time.